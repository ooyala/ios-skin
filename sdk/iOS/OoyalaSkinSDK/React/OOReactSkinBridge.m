//
//  OOReactSkinBridge.m
//  OoyalaSkinSDK
//
//  Created on 8/14/18.
//  Copyright Â© 2018 ooyala. All rights reserved.
//

#import "OOReactSkinBridge.h"
#import "OOReactSkinBridgeModule.h"
#import <React/RCTBridge+Private.h>

@interface OOReactSkinBridge ()

@property (nonatomic, weak, readonly) id<OOReactSkinBridgeDelegate> delegate;

@end


@implementation OOReactSkinBridge

@dynamic delegate;

- (instancetype)initWithDelegate:(id<OOReactSkinBridgeDelegate>)delegate
                   launchOptions:(NSDictionary *)launchOptions {
  return [super initWithDelegate:delegate launchOptions:launchOptions];
}

- (void)invalidate {
  [super invalidate];
  [NSNotificationCenter.defaultCenter removeObserver:self];
}

- (void)setUp {
  [super setUp];
  [NSNotificationCenter.defaultCenter addObserver:self
                                         selector:@selector(didInitializeModuleNotification:)
                                             name:RCTDidInitializeModuleNotification
                                           object:self.batchedBridge];
}

- (void)didInitializeModuleNotification:(NSNotification *)notification {
  id module = notification.userInfo[@"module"];

  if ([self.delegate respondsToSelector:@selector(bridge:didLoadModule:)]) {
    if (module && [module conformsToProtocol:@protocol(OOReactSkinBridgeModule)]) {
      [self.delegate bridge:self didLoadModule:module];
    }
  }
}

@end


@implementation RCTBridge (OOReactSkinEventsEmitterable)

- (OOReactSkinEventsEmitter *)skinEventsEmitter {
  return [self moduleForName:[OOReactSkinEventsEmitter moduleName]];
}

@end
