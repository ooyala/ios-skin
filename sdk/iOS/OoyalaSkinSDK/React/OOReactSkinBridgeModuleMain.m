//
//  OOReactSkinBridgeModuleMain.m
//  OoyalaSkinSDK
//
//  Created by Maksim Kupetskii on 8/14/18.
//  Copyright Â© 2018 ooyala. All rights reserved.
//

#import "OOReactSkinBridgeModuleMain.h"
#import "NSMutableDictionary+Utils.h"

#import <OoyalaSDK/OODebugMode.h>
#import <OoyalaSDK/OOOoyalaPlayer.h>

@interface OOReactSkinBridgeModuleMain ()

@property (nonatomic, weak, nullable) id<OOReactSkinModelDelegate> skinModelDelegate;

@end


@implementation OOReactSkinBridgeModuleMain

@synthesize bridge;

RCT_EXPORT_MODULE(OOReactSkinBridgeModuleMain);

- (dispatch_queue_t)methodQueue {
  return dispatch_get_main_queue();
}

- (OOReactSkinBridgeModuleType)moduleType {
  return OOReactSkinBridgeModuleTypeMain;
}

- (void)setSkinViewDelegate:(nonnull id<OOReactSkinModelDelegate>)delegate {
  _skinModelDelegate = delegate;
}

#pragma mark - Constants
#pragma mark Buttons
static NSString *playPauseButtonName    = @"PlayPause";
static NSString *playButtonName         = @"Play";
static NSString *rewindButtonName       = @"rewind";
static NSString *socialShareButtonName  = @"SocialShare";
static NSString *fullscreenButtonName   = @"fullscreen";
static NSString *learnMoreButtonName    = @"LearnMore";
static NSString *skipButtonName         = @"Skip";
static NSString *moreOptionButtonName   = @"More";
static NSString *upNextDismiss          = @"upNextDismiss";
static NSString *upNextClick            = @"upNextClick";
static NSString *adIconButtonName       = @"Icon";
static NSString *adOverlayButtonName    = @"Overlay";
static NSString *stereoscopicButtonName = @"stereoscopic";
static NSString *pipButtonName          = @"PIP";
static NSString *replayButtonName       = @"replay";

#pragma mark Keys
static NSString *nameKey              = @"name";
static NSString *embedCodeKey         = @"embedCode";
static NSString *percentageKey        = @"percentage";
static NSString *languageKey          = @"language";
static NSString *bucketInfoKey        = @"bucketInfo";
static NSString *actionKey            = @"action";
static NSString *audioTrackKey        = @"audioTrack";
static NSString *indexKey             = @"index";
static NSString *clickUrlKey          = @"clickUrl";
static NSString *eventNameKey         = @"eventName";
static NSString *playbackSpeedRateKey = @"playbackSpeedRate";
static NSString *volumeKey            = @"volume";

#pragma mark Values
static NSString *startValue   = @"start";
static NSString *moveValue    = @"move";
static NSString *endValue     = @"end";
static NSString *clickValue   = @"click";
static NSString *impressValue = @"impress";

#pragma mark - Exported Methods

RCT_EXPORT_METHOD(onVolumeChanged:(NSDictionary *)parameters) {
  float volume = [parameters[volumeKey] floatValue];
  [self.skinModelDelegate handleVolumeChanged:volume];
}

RCT_EXPORT_METHOD(onMounted) {
  LOG(@"onMounted - Not going to use at the moment");
}

RCT_EXPORT_METHOD(onLanguageSelected:(NSDictionary *)parameters) {
  [self.skinModelDelegate handleLanguageSelection:parameters[languageKey]];
}

RCT_EXPORT_METHOD(onAudioTrackSelected:(NSDictionary *)parameters) {
  [self.skinModelDelegate handleAudioTrackSelection:parameters[audioTrackKey]];
}

RCT_EXPORT_METHOD(onPlaybackSpeedRateSelected:(NSDictionary *)parameters) {
  [self.skinModelDelegate handlePlaybackSpeedRateSelection:parameters[playbackSpeedRateKey]];
}

RCT_EXPORT_METHOD(onPress:(NSDictionary *)parameters) {
  NSString *buttonName = parameters[nameKey];

  if ([buttonName isEqualToString:playPauseButtonName]) {
    [self.skinModelDelegate handlePlayPause];
  } else if([buttonName isEqualToString:playButtonName]) {
    [self.skinModelDelegate handlePlay];
  } else if([buttonName isEqualToString:replayButtonName]) {
    [self.skinModelDelegate handleReplay];
  } else if([buttonName isEqualToString:rewindButtonName]) {
    [self.skinModelDelegate handleRewind];
  } else if([buttonName isEqualToString:socialShareButtonName]) {
    [self.skinModelDelegate handleSocialShare];
  } else if([buttonName isEqualToString:fullscreenButtonName]) {
    [self.skinModelDelegate toggleFullscreen];
  } else if([buttonName isEqualToString:learnMoreButtonName]) {
    [self.skinModelDelegate handleLearnMore];
  } else if([buttonName isEqualToString:skipButtonName]) {
    [self.skinModelDelegate handleSkip];
  } else if ([buttonName isEqualToString:adIconButtonName]) {
    [self.skinModelDelegate handleIconClick:[parameters[indexKey] integerValue]];
  } else if([buttonName isEqualToString:moreOptionButtonName]) {
    [self.skinModelDelegate handleMoreOption];
  } else if([buttonName isEqualToString:upNextDismiss]) {
    [self.skinModelDelegate handleUpNextDismiss];
  } else if([buttonName isEqualToString:upNextClick]) {
    [self.skinModelDelegate handleUpNextClick];
  } else if ([buttonName isEqualToString:adOverlayButtonName]) {
    [self.skinModelDelegate handleOverlay:parameters[clickUrlKey]];
  } else if ([buttonName isEqualToString:pipButtonName]) {
    [self.skinModelDelegate handlePip];
  } else if ([buttonName isEqualToString:stereoscopicButtonName]) {
    [self.skinModelDelegate toggleStereoMode];
  }
}

RCT_EXPORT_METHOD(handleTouchStart:(NSDictionary *)params){
  NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithDictionary:params];
  [result mergeWith:@{eventNameKey: startValue}];
  [self.skinModelDelegate handleTouch:result];
}

RCT_EXPORT_METHOD(handleTouchMove:(NSDictionary *)params) {
  NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithDictionary:params];
  [result mergeWith:@{eventNameKey: moveValue}];
  [self.skinModelDelegate handleTouch:result];
}

RCT_EXPORT_METHOD(handleTouchEnd:(NSDictionary *)params) {
  NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithDictionary:params];
  [result mergeWith:@{eventNameKey: endValue}];
  [self.skinModelDelegate handleTouch:result];
}

RCT_EXPORT_METHOD(onScrub:(NSDictionary *)parameters) {
  Float64 position = [parameters[percentageKey] doubleValue];
  [self.skinModelDelegate handleScrub:position];
}

RCT_EXPORT_METHOD(setEmbedCode:(NSDictionary *)parameters) {
  NSString *embedCode = parameters[embedCodeKey];
  [self.skinModelDelegate setEmbedCode:embedCode];
}

RCT_EXPORT_METHOD(onDiscoveryRow:(NSDictionary *)parameters) {
  NSString *action     = parameters[actionKey];
  NSString *bucketInfo = parameters[bucketInfoKey];

  if ([action isEqualToString:clickValue]) {
    NSString *embedCode = parameters[embedCodeKey];
    [self.skinModelDelegate handleDiscoveryClick:bucketInfo embedCode:embedCode];
  } else if ([action isEqualToString:impressValue]) {
    [self.skinModelDelegate handleDiscoveryImpress:bucketInfo];
  }
}

@end
