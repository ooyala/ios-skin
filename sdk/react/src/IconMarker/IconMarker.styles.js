// @flow

import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  root: {
    bottom: 10,
    height: 40,
    left: 0,
    position: 'absolute',
    width: 40,
  },
});
