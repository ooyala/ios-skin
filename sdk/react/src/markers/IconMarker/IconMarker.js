// @flow

import * as React from 'react';
import { Image, TouchableWithoutFeedback } from 'react-native';
import type { ViewStyleProp } from 'react-native/Libraries/StyleSheet/StyleSheet';

import styles from './IconMarker.styles';

type Props = {
  iconUrl: ?string,
  imageUrl: ?string,
  onSeek: () => void,
  style?: ViewStyleProp,
};

type State = {
  isExpanded: boolean,
};

export default class IconMarker extends React.Component<Props, State> {
  static defaultProps = {
    style: undefined,
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      isExpanded: false,
    };

    (this: Object).handlePress = this.handlePress.bind(this);
  }

  handlePress() {
    const { imageUrl, onSeek } = this.props;
    const { isExpanded } = this.state;

    // Trigger seek callback if the marker has no image or has been expanded. If there is an image we have to show it
    // first and only after the second click on that image trigger the callback.
    if (!imageUrl || isExpanded) {
      onSeek();
    }

    this.setState({
      isExpanded: !isExpanded,
    });
  }

  render() {
    const { iconUrl, imageUrl, style } = this.props;
    const { isExpanded } = this.state;

    // If the marker is expanded and there is `imageUrl`, we have to use it.
    const image = (isExpanded && imageUrl ? imageUrl : iconUrl);

    // Image can still be absent.
    if (!image) {
      return null;
    }

    return (
      <TouchableWithoutFeedback onPress={this.handlePress}>
        <Image source={{ uri: image }} style={[styles.root, style]} />
      </TouchableWithoutFeedback>
    );
  }
}
