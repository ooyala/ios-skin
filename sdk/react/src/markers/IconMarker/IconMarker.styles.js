// @flow

import { StyleSheet } from 'react-native';

import { MARKERS_SIZES } from '../../../constants';
import commonMarkerStyle from '../styles/commonMarkerStyle';
import triangleStyle from '../styles/triangleStyle';

export default StyleSheet.create({
  expanded: {
    height: MARKERS_SIZES.ICON_EXPANDED_SIZE,
    transform: [
      { translateX: MARKERS_SIZES.ICON_EXPANDED_SIZE / -2 },
    ],
    width: MARKERS_SIZES.ICON_EXPANDED_SIZE,
  },
  image: {
    height: '100%',
    width: '100%',
  },
  root: {
    ...commonMarkerStyle,
    height: MARKERS_SIZES.ICON_COLLAPSED_SIZE,
    transform: [
      { translateX: MARKERS_SIZES.ICON_COLLAPSED_SIZE / -2 },
    ],
    width: MARKERS_SIZES.ICON_COLLAPSED_SIZE,
  },
  triangle: triangleStyle,
});
