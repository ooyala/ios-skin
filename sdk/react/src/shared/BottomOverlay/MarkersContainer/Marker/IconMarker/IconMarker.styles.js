// @flow

import { StyleSheet } from 'react-native';

import { MARKERS_SIZES } from '../../../../../constants';
import commonMarkerStyle from '../styles/commonMarker.style';
import triangleStyle from '../styles/triangle.style';

export default (leftPosition: number, containerWidth: number) => StyleSheet.create({
  expanded: {
    height: MARKERS_SIZES.ICON_EXPANDED_SIZE,
    transform: [
      { translateX: MARKERS_SIZES.ICON_EXPANDED_SIZE / -2 },
    ],
    width: MARKERS_SIZES.ICON_EXPANDED_SIZE,
  },
  image: {
    height: '100%',
    width: '100%',
  },
  root: {
    ...commonMarkerStyle,
    height: MARKERS_SIZES.ICON_COLLAPSED_SIZE,
    left: leftPosition,
    transform: [
      { translateX: MARKERS_SIZES.ICON_COLLAPSED_SIZE / -2 },
    ],
    width: MARKERS_SIZES.ICON_COLLAPSED_SIZE,
  },
  triangle: triangleStyle,
});
