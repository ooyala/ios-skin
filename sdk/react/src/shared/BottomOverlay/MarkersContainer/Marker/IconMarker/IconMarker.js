// @flow

import React from 'react';
import { Image, TouchableWithoutFeedback, View } from 'react-native';
import type { ViewStyleProp } from 'react-native/Libraries/StyleSheet/StyleSheet';

import createStyles from './IconMarker.styles';

type Props = {
  backgroundColor?: string,
  containerWidth: number,
  iconUrl?: string,
  imageUrl?: string,
  leftPosition: number,
  onSeek: () => void,
  onTouch: () => void,
  style?: ViewStyleProp,
  touchColor?: string,
};

type State = {
  isExpanded: boolean,
};

export default class IconMarker extends React.Component<Props, State> {
  static defaultProps = {
    backgroundColor: undefined,
    iconUrl: undefined,
    imageUrl: undefined,
    style: undefined,
    touchColor: undefined,
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      isExpanded: false,
    };

    (this: Object).handlePress = this.handlePress.bind(this);
  }

  handlePress() {
    this.setState(({ isExpanded }, { imageUrl, onSeek, onTouch }) => {
      onTouch();

      // Trigger seek callback if the marker has no image or has been expanded. If there is an image we have to show it
      // first and only after the second click on that image trigger the callback.
      if (!imageUrl || isExpanded) {
        onSeek();
      }

      return {
        isExpanded: !isExpanded,
      };
    });
  }

  render() {
    const {
      backgroundColor, containerWidth, iconUrl, imageUrl, leftPosition, style, touchColor,
    } = this.props;
    const { isExpanded } = this.state;

    const styles = createStyles(leftPosition, containerWidth);

    // If the marker is expanded and there is `imageUrl`, we have to use it.
    const image = (isExpanded && imageUrl ? imageUrl : iconUrl);

    // Image can still be absent.
    if (!image) {
      return null;
    }

    const expandedStyles = [];

    // Apply expanded style only if the larger image is present.
    if (isExpanded && imageUrl) {
      expandedStyles.push(styles.expanded);

      if (touchColor) {
        expandedStyles.push({ backgroundColor: touchColor });
      }
    }

    return (
      <TouchableWithoutFeedback onPress={this.handlePress}>
        <View
          style={[
            styles.root,
            style,
            backgroundColor && { backgroundColor },
            ...expandedStyles,
          ]}
        >
          <Image source={{ uri: image }} style={styles.image} />
          <View
            style={[
              styles.triangle,
              // Apply expanded style only if the larger image is present.
              isExpanded && imageUrl && touchColor && { borderTopColor: touchColor },
              isExpanded && imageUrl && styles.triangleExpanded,
            ]}
          />
        </View>
      </TouchableWithoutFeedback>
    );
  }
}
