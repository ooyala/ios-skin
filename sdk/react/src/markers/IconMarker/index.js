// @flow

import IconMarker from './IconMarker';

export default IconMarker;
