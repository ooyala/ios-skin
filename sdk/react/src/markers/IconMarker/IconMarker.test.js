// @flow

import React from 'react';
import TestRenderer from 'react-test-renderer';

import IconMarker from './IconMarker';

describe('IconMarker', () => {
  it('renders matching snapshot', () => {
    const testRenderer = TestRenderer.create(
      <IconMarker
        iconUrl="http://example.com/icon.png"
        imageUrl="http://example.com/image.png"
        onSeek={() => undefined}
      />,
    );

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('does not render anything when URLs are empty', () => {
    const testRenderer = TestRenderer.create(
      <IconMarker
        iconUrl={undefined}
        imageUrl={undefined}
        onSeek={() => undefined}
      />,
    );

    expect(testRenderer.toJSON()).toBeNull();
  });
});
