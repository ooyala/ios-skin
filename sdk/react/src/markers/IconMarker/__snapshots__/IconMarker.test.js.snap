// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconMarker renders matching snapshot 1`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "black",
        "borderColor": "#4E4E4E",
        "borderRadius": 2,
        "borderWidth": 1,
        "bottom": 6,
        "height": 32,
        "padding": 2,
        "position": "absolute",
        "shadowColor": "black",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.5,
        "shadowRadius": 4,
        "transform": Array [
          Object {
            "translateX": -16,
          },
        ],
        "width": 32,
      },
      undefined,
      undefined,
    ]
  }
>
  <Image
    source={
      Object {
        "uri": "http://example.com/icon.png",
      }
    }
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  />
  <View
    style={
      Array [
        Object {
          "borderBottomWidth": 0,
          "borderColor": "transparent",
          "borderTopColor": "#4E4E4E",
          "borderWidth": 6,
          "bottom": -6,
          "height": 0,
          "left": "50%",
          "position": "absolute",
          "transform": Array [
            Object {
              "translateX": -3,
            },
          ],
          "width": 0,
        },
        false,
      ]
    }
  />
</View>
`;
